oslabs-beta/watchmo
# multi-project CLI functionality

This is a big one! In this pull request we completely refactored the CLI to make it more modular and to support multiple projects. Now, in theory, a user can globally install watchmo and point at any project they want. 

**How to read this request**
There are three sections:
- Part I: What stayed the same
- Part II: What's new
- Part III: How to test it
Part 1 is required reading. Follow Part 2 to make sure everything is working. Part 3 is for reference, and understanding how it works. If you just want to start working, you can skim Part 3, otherwise, I'd recommend reading through Part 3 to get an understanding of the changes, and then run through the tests in Part 2 to see it in action.

**Part I: What stayed the same**
All the previous commands, `watch`, `mo`, and `less`,  still work the way they did before when given no argument, except with a slight change to the file path, saving all data to /watchmoData/default/... rather than directly to /watchmoData/.... The frontend engineers will need to reconfigure how this data is being fetched because of this file restructuring. 

**Part II: How to test it**
As stated in Part I, if you want, you don't have to change anything and it should still work, except for the changed paths from the server (see Part III: "3. The server paths are different..."). However, to test the new functionality, do the following:

_adding new projects_
1. Run `watchmo config projectName -d`. This should create a new file /watchmoData/projectName, and a pre-written config.json file that matches the development config file pointing towards Evan's DummyGraphQL server we have been using this whole time. It should also add the name "projectName" to the array in /watchmoData/projectNames.json. (_Note_: if you don't include the -d tag, you will get an unconfigured config.json file)

_using the new positional arguments_
1. Make sure the Evan's GQL server is running, then run `watchmo watch projectName`.
2. Wait for data to populate in /watchmoData/projectName/snapshots.txt. Stop watchmo watch (with `^C` for instance) and run `watchmo mo projectName`. This should create a parsedData file, /watchmoData/projectName/parsedData.json.

_testing the server routes_
1. Run `watchmo mo projectName -o -b`. This should open the server, stop the data from rewriting, and open up the browser. The browser should _not_ display any data until the paths have been reconfigured.
2. Open up postman. Test the following routes:
    -`htttp://localhost:3333/projectName/config.json`: should return the config.json file
    - `htttp://localhost:3333/projectName/parsedData.json`: should return the parsedData.json file
    - `htttp://localhost:3333/projectNames.json`: should return an array, ["default", "projectName"]
3. Close the watchmo server. 

_cleaning up files_
1. Run `watchmo less projectName`. This should clear the files snapshots.txt and parsedData.json in the /watchmoData/projectName directory.
2. Run `watchmo less projectName --remove`. This should delete the project directory in /watchmoData, and remove the name "projectName" from the array in /watchmoData/projectNames.json. (_Note_: This command can be run without cleaning the files as done in step 7, they are both included to test that they are both working.)

**What's new**
1. A new command, _configure_! Suppose your new project is named 'foo'. Then:
    - `watchmo configure foo` will create a new project directory /foo in /watchmoData, and create a config.json file in /watchmoData/foo.
    - _For Developers!_ `watchmo configure foo --development //alias -d`: Creates the project and the config file, but copies the config file from default, so you don't have to manually change the config.json file for each new project you want to test.
2. Positional arguments in yargs allowing watchmo to point any project for any command. More explicitly, for a project named 'foo':
    - `watchmo watch foo` makes queries to the endpoint, both specified in /watchmoData/foo/config.json
    - `watchmo less foo` cleans up the `/snapshots.txt` and /parsedData.json files in /watchmoData/foo/.
    - `watchmo mo foo` parses the data in /watchmoData/foo/snapshots.txt and saves it to /watchmoData/foo/parsedData.json. The `watchmo mo --open //alias -o` and 
`watchmo mo --bundle //alias -b` flags still work as before, opening the server and suppressing the data-parsing, respectively.
3. New flag for less. `watchmo less foo --remove //alias -r` will delete the entire /foo directory
2. The /watchmoData file structure now includes a JSON list of all the project names, and a different directory for each project:
```
/*
 Used to be this...

/watchmoData
|
|-> config.json
|-> parsedData.json
|-> snapshots.txt

... Is now this 

/watchmoData
|
|-> projectNames.json (This contains an array of names matching the directories below)
|
|-> /default
|     |
|     |-> config.json
|     |-> parsedData.json
|     |-> snapshots.txt
|
|-> /[another project name]
|     |
|     |-> config.json
|     |-> parsedData.json
|     |-> snapshots.txt
|
... and continues for each new project
*/
```

3. The server paths are different. The old paths are still maintained, but they are deprecated and should ultimately be removed. Instead, everything is now statically served through /watchmoData.
Therefore...
    - To get parsed data from a project named 'foo', use `/foo/parsedData.json`
    - To get the config file from a project named 'foo',  use `/foo/config.json`
    - To get the project names currently in watchmo's sight, use `/foo/projectNames.json`

4. Codebase redesign for the commands using utility functions. In /src/commands, a new directory was created, /utility, currently with three files: serverHelpers.js, pathHelpers.js, and fileHelpers.js. By separating out the logic for file manipulation, server manipulation, and path manipulation, the actual command files are easier to follow. 

    Most of these functions are straightforward helper functions, but one design decision of note is dataPaths. calling `dataPaths(foo)` returns an object with many paths to many directories, including:
    - path to the project file .../watchmoData/foo
    - path to the config files .../watchmoData/foo/config.json
    - path to the parsed and raw data files .../watchmoData/foo/(parsedData.json || rawData.json)
    - some static paths to the config.json template and the projectNames.json file
By doing this, adding a new command with a positional argument of [project] can easily point to all the necessary directories, removing the need for cluttered path.join statements and confusing navigation.

    This process is not complete, pathHelpers is currently empty and less still references the fs module. A good rule of thumb for this design choice is that no command file will reference an external module. 
**NOTE** serverHelpers.js now contains the PORT variable which server.js uses. This external referencing was necessary to make the openServer function modular with respect to the port.
